crates/turborepo-lib/src/daemon/bump_timeout.rs

```diff
@@ -0,0 +1,60 @@
+use std::{
+    sync::atomic::{AtomicU64, Ordering},
+    time::Duration,
+};
+
+use tokio::time::Instant;
+
+/// A timeout that can be bumped forward in time by calling reset.
+///
+/// Calling reset with a new duration will change the deadline
+/// to the current time plus the new duration. It is non-mutating
+/// and can be called from multiple threads.
+#[derive(Debug)]
+pub struct BumpTimeout(Instant, AtomicU64);
+
+impl BumpTimeout {
+    pub fn new(duration: Duration) -> Self {
+        let start = Instant::now();
+        let millis = duration.as_millis();
+        Self(start, AtomicU64::new(millis as u64))
+    }
+
+    pub fn duration(&self) -> Duration {
+        Duration::from_millis(self.1.load(Ordering::Relaxed))
+    }
+
+    pub fn deadline(&self) -> Instant {
+        self.0 + self.duration()
+    }
+
+    pub fn elapsed(&self) -> Duration {
+        self.0.elapsed()
+    }
+
+    /// Resets the deadline to the current time plus the given duration.
+    pub fn reset(&self, duration: Duration) {
```

Comment:
Do you mind elaborating a little bit. We don't do any read-then-writes, so maybe I am missing something.

Reply:
Let's say we have three threads. Thread A passes 5s and thread B passes 1s. Thread C calls wait. This sequence of events could happen (I think)

A -> sets duration to 5s + elapsed
B -> sets duration to 1s + elapsed
A -> stores duration as 5s + elapsed
C -> calls wait reads duration as 5s + elapsed, sleeps until then
B -> stores duration as 1s + elapsed

I'm not sure this is a big issue because really it just means a thread may sleep for a little too long.